scheduler/src/cook/mesos/util.clj

```diff
@@ -24,44 +24,58 @@
             [metatransaction.core :refer (db)]
             [metrics.timers :as timers]
             [plumbing.core :as pc :refer (map-vals map-keys)])
-  (:import [java.util Date]))
+  (:import [java.util Date]
+           [com.google.common.cache CacheBuilder]
+           [java.util.concurrent TimeUnit]))
+
+
+(defn new-cache []
+  "Build a new cache"
+  (-> (CacheBuilder/newBuilder)
+      (.maximumSize 1000000)
+      ;; if its not been accessed in 2 hours, whatever is going on, its not being visted by the
+      ;; scheduler loop anymore. E.g., its probably failed/done and won't be needed. So,
+      ;; lets kick it out to keep cache small.
+      (.expireAfterAccess 2 TimeUnit/HOURS)
+      (.build)))
 
 (defn lookup-cache-generic!
-      "Generic cache. Caches under a key (extracted from the item with extract-key-fn. Uses miss-fn to fill
-      any misses. Caches only postive hits where both functions return non-nil"
-      [cache extract-key-fn miss-fn item]
-      (let [key (extract-key-fn item)]
-           (if key
-             (cache/lookup (swap! cache
-                                  #(if (cache/has? % key)
-                                     (cache/hit % key)
-                                     (let [result (miss-fn item)]
-                                          ; Only cache non-nil
-                                          (if result
-                                            (cache/miss % key result)
-                                            %))))
-                           key)
-             (miss-fn item))))
+  "Generic cache. Caches under a key (extracted from the item with extract-key-fn. Uses miss-fn to fill
+  any misses. Caches only postive hits where both functions return non-nil"
+  [cache-atom extract-key-fn miss-fn item]
+  (let [key (extract-key-fn item)
+        cache (deref cache-atom)]
+    (if key
+      (let [result (.getIfPresent cache key)]
+        (if result
+          result ; we got a hit.
+          (let [new-result (miss-fn item)]
+            ; Only cache non-nil
+            (if new-result
```

Comment:
Can we use `if-let` here?

Reply:
Not obviously. We want to unconditionally return new-result, so it needs to leave the scope of the 'if'.